### PR TITLE
Hooks support

### PR DESCRIPTION
**Pull request by delvedor**
As discussed in #44 this pr adds the support for hooks.
At the moment we support `onRequest` and `onRequestRaw`, in the first one *request* and *reply* are from fastify are passed as parameters, in the second one the *request* and *response* node core objects.
`onRequestRaw` is always called before `onRequest`.

Usage:
```js
fastify.addHook({
  onRequest: function (request, reply, next) {
    // stuff
    next()
  }
})
```

Thoughts?

### CODE REVIEW

File: fastify.js [hunk 1]
```diff
@@ -2,16 +2,19 @@
...
 const wayfarer = require('wayfarer')
 const stripUrl = require('pathname-match')
-const pluginLoader = require('boot-in-the-arse')
+const pluginLoader = require('avvio')
 const http = require('http')
 const https = require('https')
 const pinoHttp = require('pino-http')
 const Middie = require('middie')
-const Reply = require('./lib/reply')
+const mapSeries = require('steed')().mapSeries
```
Comment (mcollina) [inline fastify.js]: we should probably want to do this inside `build()`, to limit it to a fastify instance.

File: lib/hooks.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
+'use strict'
+
+const hooks = {
+  onRequest: [],
+  onRequestRaw: []
+}
```
Comment (mcollina) [inline lib/hooks.js]: is this a global? It seems all this object should be built for each fastify instance.

File: lib/tier-node.js [hunk 3]
```diff
@@ -89,9 +89,24 @@ function handler (handle, params, req, res, body, query) {
...
   const reply = new Reply(req, res, handle)
-  const result = handle.handler(request, reply)
-  if (result && typeof result.then === 'function') {
-    reply.send(result)
+
+  mapSeries(onRequest(), _onRequestIterator, _runOnRequestWrap(request, reply, handle))
+
+  function _onRequestIterator (fn, cb) {
+    setImmediate(fn, request, reply, cb)
+  }
```
Comment (mcollina) [inline lib/tier-node.js]: I do not think we should allocate this new closure here, if possible.

File: lib/tier-node.js [hunk 4]
```diff
@@ -89,9 +89,24 @@ function handler (handle, params, req, res, body, query) {
...
+function _runOnRequestWrap (request, reply, handle) {
+  return function _runOnRequest (err) {
+    if (err) {
+      reply.send(err)
+      return
+    }
+    const result = handle.handler(request, reply)
+    if (result && typeof result.then === 'function') {
+      reply.send(result)
+    }
```
Comment (mcollina) [inline lib/tier-node.js]: same here, see https://github.com/mcollina/aedes/blob/master/lib/handlers/subscribe.js#L25-L29 to avoid creating closures.

**Comment by delvedor** [comment]
With the above commit the hooks are limited to the fastify instance, but we must pass them as parameter along all tier-node.

File: test/internals/reply.test.js [hunk 5]
```diff
@@ -22,7 +22,7 @@ test('Once called, Reply should return an object with methods', t => {
   t.is(typeof reply.send, 'function')
   t.is(typeof reply.code, 'function')
   t.is(typeof reply.header, 'function')
-  t.strictEqual(reply.req, request)
+  t.strictEqual(reply._req, request)
```
Comment (mcollina) [inline test/internals/reply.test.js]: why this change? I think `reply.req` is quite handy and it should be part of our public API.

File: fastify.js [hunk 6]
```diff
@@ -70,7 +83,31 @@ function build (options) {
...
+    )
+  }
+
+  function State (req, res) {
+    this.req = req
+    this.res = res
+  }
+
+  function _onRequestIterator (fn, cb) {
+    setImmediate(fn, this.req, this.res, cb)
```
Comment (mcollina) [inline fastify.js]: why `setImmediate()`? Why can't we call the function directly?

File: test/internals/reply.test.js (hunk 5 above)
Comment (delvedor) [inline test/internals/reply.test.js]: The user can access the req from `request.req`, inside `reply` `req` is needed just for logging the errors :)

Comment (mcollina) [inline test/internals/reply.test.js]: ok!

File: fastify.js (hunk 6 above)
Comment (delvedor) [inline fastify.js]: Because we have seen that when performing a lot of async operations, call the callbacks with `setImmediate` helps to improve the performances.
Do you think this is not the case?

**Comment by delvedor** [comment]
I'm really happy to say that if we run the benchmarks without adding hooks, the benchmarks aren't affected! :tada:

File: fastify.js (hunk 6 above)
Comment (mcollina) [inline fastify.js]: Yes, but this adds a `setImmediate` for every hook executed. I think we should avoid this, and just retain the only `setImmediate` we have later on.

**Comment by delvedor** [comment]
Before commit the docs we must decide which hooks support with this pr.

In the related issue, @davidmarkclements proposed:
*preMiddleware*
*postMiddleware*
*preValidation/postRouter*
*postValidation/preReply*

At the moment we are supporting *onRequestRaw* and *onRequest*.

**Comment by mcollina** [comment]
I think we should start simple with a couple of those, and add them if they are needed. I like better the "pre"-"post" approach, how about:

postRouter
preHandler

But I do not care much about how we name those. What do you think works best?

**Comment by davidmarkclements** [comment]
preValidation > postRouter imo - people focus on and understand validation must faster, and you see the use cases more clearly than postRouter 

(you could also have both and alias, so they're the same thing)

**Comment by lucamaraschi** [comment]
I am just concerned about the DX and how the different triggers might be perceived by the developers. I agree with @davidmarkclements that validation is going to be more "immediate" to understand...can we find an easy to digest alias also for `preHandler`?

**Comment by delvedor** [comment]
I like the idea of the aliases, since the internal architecture of fastify is linear we could alias most of the hooks.

```
NEW REQUEST => preMiddleware => MIDDLEWARE => postMiddleware/preRouting =>
=> ROUTING => postRouting/preParsing => PARSING => postParsing/preValidation =>
=> VALIDATION => postValidation/preHandler => HANDLER
```

In this way the user can choose the most appropriate name for his logic.

**Comment by mcollina** [comment]
On the other side, aliasing is confusing. Let's stop bikeshedding and just do one!

**Comment by delvedor** [comment]
I tried to implement all the hooks listed above, using the aliasing for all the *post/pre*, as you can see in the code.

I runned the benchmarks and I noticed a regression of ~5% when *no hook* is registered, if I register *all the hooks* with just `next()` I get a ~8% regression.

I think these are all the hooks we can support, but I don't know if all of them are really useful (like *postParsing/preValidation* for example).

**Comment by mcollina** [comment]
If we remove preValidation, how much we gain?

**Comment by delvedor** [comment]
-4% with no hooks registered
-7% with all hooks with just `next()`

File: lib/hooks.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+  function _add (hook, fn) {
+    if (supportedHooks.indexOf(hook) === -1) {
+      throw new Error(`${hook} hook not supported!`)
+    }
+
+    if (hook === 'postMiddleware') hook = 'preRouting'
+    if (hook === 'postRouting') hook = 'preParsing'
+    if (hook === 'postParsing') hook = 'preValidation'
+    if (hook === 'postValidation') hook = 'preHandler'
```
Comment (mcollina) [inline lib/hooks.js]: you can implement this with an alias table.

**Comment by mcollina** [comment]
can we remove another one?

**Comment by delvedor** [comment]
Sure, the other one that in my opinion is not that useful is *postRouting/preParsing*.

In this way we will have:
```
preMiddleware => preRouting => preHanlder
```
And I think it could be a good idea rename *preMiddleware* to *onRequest*.

File: README.md [hunk 8]
```diff
@@ -401,6 +402,27 @@ fastify.listen(3000, err => {
...
+fastify.addHook({
+  onRequest: function (request, response, next) {
+    // perform some operation
+    next()
+  }
+})
+```
+
+`addHook` can accept an array of hooks, and you can register more than one callback for every hook.  
+*preHandler* gets as parameters the same object of the route handler, the fastify `request` and `reply` objects.
```
Comment (mcollina) [inline README.md]: are we sure this is the easier API for this?

How about:

```js
fastify.addHook('onRequest', function (request, response, next) {
})
```

**Comment by delvedor** [comment]
What about `fastify.subscribeHook()`?

File: README.md (hunk 8 above)
Comment (delvedor) [inline README.md]: I like it, but we lose the "add multiple hooks as an array", is it ok for you?

Comment (mcollina) [inline README.md]: That's ok for me.

**Comment by mcollina** [comment]
addHook is better

**Comment by delvedor** [comment]
Done!
I also made `addHook` chainable :)

**Comment by lucamaraschi** [comment]
LGTM

**Comment by delvedor** [comment]
Merged, thank you all! :)
